src/threading/atomic.h

```diff
@@ -29,64 +29,83 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #define GCC_VERSION   (__GNUC__        * 100 + __GNUC_MINOR__)
 #define CLANG_VERSION (__clang_major__ * 100 + __clang_minor__)
 #if GCC_VERSION >= 407 || CLANG_VERSION >= 302
-	#define ATOMIC_LOAD(T, v)      return __atomic_load_n(&(v),       __ATOMIC_SEQ_CST)
-	#define ATOMIC_STORE(T, v, x)  __atomic_store (&(v), &(x), __ATOMIC_SEQ_CST); return x
-	#define ATOMIC_ADD_EQ(T, v, x) return __atomic_add_fetch(&(v), (x), __ATOMIC_SEQ_CST)
-	#define ATOMIC_SUB_EQ(T, v, x) return __atomic_sub_fetch(&(v), (x), __ATOMIC_SEQ_CST)
-	#define ATOMIC_POST_INC(T, v)  return __atomic_fetch_add(&(v), 1, __ATOMIC_SEQ_CST)
-	#define ATOMIC_POST_DEC(T, v)  return __atomic_fetch_sub(&(v), 1, __ATOMIC_SEQ_CST)
+	#define ATOMIC_LOAD(T, v)        return __atomic_load_n    (&(v),       __ATOMIC_SEQ_CST)
+	#define ATOMIC_STORE(T, v, x)           __atomic_store     (&(v), &(x), __ATOMIC_SEQ_CST); return x
+	#define ATOMIC_EXCHANGE(T, v, x) return __atomic_exchange_n(&(v), (x),  __ATOMIC_SEQ_CST)
+	#define ATOMIC_ADD_EQ(T, v, x)   return __atomic_add_fetch (&(v), (x),  __ATOMIC_SEQ_CST)
+	#define ATOMIC_SUB_EQ(T, v, x)   return __atomic_sub_fetch (&(v), (x),  __ATOMIC_SEQ_CST)
+	#define ATOMIC_POST_INC(T, v)    return __atomic_fetch_add (&(v), 1,    __ATOMIC_SEQ_CST)
+	#define ATOMIC_POST_DEC(T, v)    return __atomic_fetch_sub (&(v), 1,    __ATOMIC_SEQ_CST)
+	#define ATOMIC_CAS(T, v, e, d)   return __atomic_compare_exchange_n(&(v), &(e), (d), \
+		false, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST)
 #else
 	#define ATOMIC_USE_LOCK
 	#include "threading/mutex.h"
 
 	#define ATOMIC_LOCK_OP(T, op) do { \
-			mutex.lock(); \
-			T _val = (op); \
-			mutex.unlock(); \
-			return _val; \
+			m_mutex.lock();            \
+			T _val = (op);             \
+			m_mutex.unlock();          \
+			return _val;               \
 		} while (0)
-	#define ATOMIC_LOAD(T, v) \
+	#define ATOMIC_LOCK_CAS(T, v, e, d) do { \
+			m_mutex.lock();                  \
+			bool _eq = (v == e);             \
+			if (eq)                          \
+				v = d;                       \
+			m_mutex.unlock();                \
+			return eq;                       \
+		} while (0)
+	#define ATOMIC_LOAD(T, v)                     \
 		if (sizeof(T) <= sizeof(void*)) return v; \
 		else ATOMIC_LOCK_OP(T, v);
-	#define ATOMIC_STORE(T, v, x) \
+	#define ATOMIC_STORE(T, v, x)                     \
```

Comment:
It does if looked at with a tab width of 4: https://github.com/minetest/minetest/pull/3333/files?ts=4